File: app/src/main/java/ru/samtakoy/core/business/impl/CoursesRepositoryImpl.java
package ru.samtakoy.core.business.impl;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;
import ru.samtakoy.core.business.CoursesRepository;
import ru.samtakoy.core.business.TempCourseRepository;
import ru.samtakoy.core.database.room.MyRoomDb;
import ru.samtakoy.core.database.room.entities.LearnCourseEntity;
import ru.samtakoy.core.database.room.entities.types.CourseType;
import ru.samtakoy.core.database.room.entities.types.LearnCourseMode;
import ru.samtakoy.core.presentation.log.MyLog;

public class CoursesRepositoryImpl implements CoursesRepository {

    private static final String TAG = "CoursesRepositoryImpl";

    @Inject
    MyRoomDb db;
    @Inject
    TempCourseRepository mTempCourseRepository;

    @Inject
    public CoursesRepositoryImpl() {
    }

    /*
    public LearnCourseEntity getCourse(Long learnCourseId) {

        if (learnCourseId == mTempCourseRepository.getTempCourseId()) {
            return mTempCourseRepository.getTempCourse();
        }
        return db.courseDao().getLearnCourse(learnCourseId);
    }*/

    @Override
    @NotNull
    public Single<LearnCourseEntity> getCourseRx(@NotNull Long learnCourseId) {


        if (learnCourseId == mTempCourseRepository.getTempCourseId()) {
            return Single.fromCallable(
                    () -> mTempCourseRepository.getTempCourse()
            );
        }

        return db.courseDao().getLearnCourseRx(learnCourseId);
    }

    @Override
    public LearnCourseEntity getTempCourseFor(Long qPackId, List<Long> cardIds, Boolean shuffleCards) {
        return mTempCourseRepository.getTempCourseFor(qPackId, cardIds, shuffleCards);
    }

    @Override
    public LearnCourseEntity getCourseByMode(LearnCourseMode mode) {
        return db.courseDao().getLearnCourseByMode(mode.getDbId());
    }

    @Override
    public boolean updateCourse(LearnCourseEntity course) {

        if (course.getCourseType() == CourseType.TEMPORARY) {
            mTempCourseRepository.updateTempCourse(course);
            return true;
        }

        return db.courseDao().updateCourse(course) > 0;
    }

    @Override
    public Completable deleteCourse(long courseId) {
        return Completable.fromCallable(() -> db.courseDao().deleteCourseById(courseId) > 0);
    }

    @Override
    public Completable deleteQPackCourses(Long qPackId) {
        return Completable.fromCallable(() -> {
            db.courseDao().deleteQPackCourses(qPackId);
            return true;
        });
    }

    @Override
    public LearnCourseEntity addNewCourseNow(LearnCourseEntity newCourse) {

        if (newCourse.getCourseType() == CourseType.TEMPORARY) {
            MyLog.add(TAG + ", CourseType.TEMPORARY cant be added to Database");
            return null;
        }

        Long id = db.courseDao().addLearnCourse(newCourse);
        newCourse.setId(id);
        return newCourse;
    }

    @Override
    public Single<LearnCourseEntity> addNewCourse(LearnCourseEntity newCourse) {
        return Single.fromCallable(
                () -> {

                    if (newCourse.getCourseType() == CourseType.TEMPORARY) {
                        String error = TAG + ", CourseType.TEMPORARY cant be added to Database";
                        MyLog.add(error);
                        throw new Exception(error);
                    }

                    Long id = db.courseDao().addLearnCourse(newCourse);
                    newCourse.setId(id);
                    return newCourse;
                }
        );
    }


    @Override
    public Flowable<List<LearnCourseEntity>> getAllCourses() {
        return db.courseDao().getAllCourses();
    }

    @Override
    public Single<List<LearnCourseEntity>> getAllCoursesSingle() {
        return db.courseDao().getAllCoursesSingle();
    }

    @Override
    public Flowable<List<LearnCourseEntity>> getCoursesByIds(Long[] targetCourseIds) {
        return db.courseDao().getCoursesByIds(Arrays.asList(targetCourseIds));
    }

    private List<Integer> modesToIds(List<LearnCourseMode> targetModes) {
        List<Integer> result = new ArrayList<>(targetModes.size());
        for (LearnCourseMode mode : targetModes) {
            result.add(mode.getDbId());
        }
        return result;
    }

    @Override
    public Flowable<List<LearnCourseEntity>> getCoursesByModes(List<LearnCourseMode> targetModes) {
        return db.courseDao().getCoursesByModes(modesToIds(targetModes));
    }

    @Override
    public Flowable<List<LearnCourseEntity>> getCoursesByModes(LearnCourseMode... mode) {
        return db.courseDao().getLearnCourseByModes(Arrays.asList(mode));
    }

    @Override
    public List<LearnCourseEntity> getCoursesByModesNow(LearnCourseMode... mode) {
        return db.courseDao().getLearnCourseByModesNow(Arrays.asList(mode));
    }

    @Override
    public Flowable<List<LearnCourseEntity>> getCoursesForQPack(Long qPackId) {
        return db.courseDao().getCoursesForQPack(qPackId);
    }

    // не в Rx стиле для сервиса, перенести в отдельный репозиторий?

    @Override
    public List<LearnCourseEntity> getOrderedCoursesLessThan(LearnCourseMode mode, Date repeatDate) {
        return db.courseDao().getOrderedCoursesLessThan(mode, repeatDate);
    }

    @Override
    public List<LearnCourseEntity> getOrderedCoursesMoreThan(LearnCourseMode mode, Date repeatDate) {
        return db.courseDao().getOrderedCoursesMoreThan(mode, repeatDate);
    }
}

File: app/src/main/java/ru/samtakoy/core/data/local/reps/impl/CoursesRepositoryImpl.kt
package ru.samtakoy.core.data.local.reps.impl

import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import kotlinx.coroutines.flow.Flow
import ru.samtakoy.core.data.local.database.room.MyRoomDb
import ru.samtakoy.core.data.local.database.room.entities.LearnCourseEntity
import ru.samtakoy.core.data.local.database.room.entities.types.LearnCourseMode
import ru.samtakoy.core.data.local.reps.CoursesRepository
import java.util.Arrays
import java.util.Date
import javax.inject.Inject

class CoursesRepositoryImpl @Inject constructor(
    private val db: MyRoomDb
) : CoursesRepository {

    override suspend fun getCourse(learnCourseId: Long): LearnCourseEntity? {
        return db.courseDao().getLearnCourse(learnCourseId)
    }

    override fun getCourseAsFlow(courseId: Long): Flow<LearnCourseEntity?> {
        return db.courseDao().getLearnCourseAsFlow(courseId)
    }

    override fun getCourseFlowableRx(learnCourseId: Long): Flowable<LearnCourseEntity> {
        return db.courseDao().getLearnCourseFlowableRx(learnCourseId)
    }

    override fun getCourseByMode(mode: LearnCourseMode): LearnCourseEntity {
        return db.courseDao().getLearnCourseByMode(mode.dbId)!!
    }

    override suspend fun updateCourse(course: LearnCourseEntity): Boolean {
        return db.courseDao().updateCourse(course) > 0
    }

    override suspend fun deleteCourse(courseId: Long) {
        db.courseDao().deleteCourseById(courseId)
    }

    override suspend fun deleteQPackCourses(qPackId: Long) {
        db.courseDao().deleteQPackCourses(qPackId)
    }

    override fun addNewCourseNow(newCourse: LearnCourseEntity): LearnCourseEntity {
        val id = db.courseDao().addLearnCourse(newCourse)
        return newCourse.copy(id = id)
    }

    override suspend fun addNewCourse(newCourse: LearnCourseEntity): LearnCourseEntity {
        val id = db.courseDao().addLearnCourse(newCourse)
        return newCourse.copy(id = id)
    }

    override fun getAllCoursesRx(): Flowable<List<LearnCourseEntity>> {
        return db.courseDao().getAllCoursesRx()
    }

    override fun getAllCoursesAsFlow(): Flow<List<LearnCourseEntity>> {
        return db.courseDao().getAllCoursesAsFlow()
    }

    override suspend fun getAllCourses(): List<LearnCourseEntity> {
        return db.courseDao().getAllCourses()
    }

    override fun getCoursesByIds(targetCourseIds: Array<Long>): Flowable<List<LearnCourseEntity>> {
        return db.courseDao().getCoursesByIds(Arrays.asList(*targetCourseIds))
    }

    override fun getCoursesByIdsAsFlow(targetCourseIds: Array<Long>): Flow<List<LearnCourseEntity>> {
        return db.courseDao().getCoursesByIdsAsFlow(Arrays.asList(*targetCourseIds))
    }

    private fun modesToIds(targetModes: List<LearnCourseMode>): List<Int> {
        val result: MutableList<Int> = ArrayList(targetModes.size)
        for (mode in targetModes) {
            result.add(mode.dbId)
        }
        return result
    }

    override fun getCoursesByModes(targetModes: List<LearnCourseMode>): Flowable<List<LearnCourseEntity>> {
        return db.courseDao().getCoursesByModes(modesToIds(targetModes))
    }

    override fun getCoursesByModesAsFlow(targetModes: List<LearnCourseMode>): Flow<List<LearnCourseEntity>> {
        return db.courseDao().getCoursesByModesAsFlow(modesToIds(targetModes))
    }

    override fun getCoursesByModes(vararg mode: LearnCourseMode): Flowable<List<LearnCourseEntity>> {
        return db.courseDao().getLearnCourseByModes(Arrays.asList(*mode))
    }

    override fun getCoursesByModesNow(vararg mode: LearnCourseMode): List<LearnCourseEntity> {
        return db.courseDao().getLearnCourseByModesNow(Arrays.asList(*mode))
    }

    override fun getCoursesForQPackRx(qPackId: Long): Flowable<List<LearnCourseEntity>> {
        return db.courseDao().getCoursesForQPackRx(qPackId)
    }

    override fun getCoursesForQPackAsFlow(qPackId: Long): Flow<List<LearnCourseEntity>> {
        return db.courseDao().getCoursesForQPackAsFlow(qPackId)
    }

    // не в Rx стиле для сервиса, перенести в отдельный репозиторий?
    override fun getOrderedCoursesLessThan(mode: LearnCourseMode, repeatDate: Date): List<LearnCourseEntity> {
        return db.courseDao().getOrderedCoursesLessThan(mode, repeatDate)
    }

    override fun getOrderedCoursesMoreThan(mode: LearnCourseMode, repeatDate: Date): List<LearnCourseEntity> {
        return db.courseDao().getOrderedCoursesMoreThan(mode, repeatDate)
    }

    companion object {
        private const val TAG = "CoursesRepositoryImpl"
    }
}
